### PR TITLE
Homework #2 / tests

### PR DESCRIPTION
**Pull request by delveper**
This PR is related to the addition of unit tests, integration tests, and end-to-end tests in codebase according to Homework #2 / tests.
List of changes:
Added unit tests for all major components and services.
Added integration tests to repository.
Added E2E test to mimic the user behavior and to check if the application is working as expected in a simulated production environment using docker and postman collection.

### CODE REVIEW

File: internal/rate/handler_test.go
Comment (lumenalux) [inline internal/rate/handler_test.go:1]: Lines `65-67`

```Go
handler := http.HandlerFunc(h.Rate)
handler.ServeHTTP(rw, req)
```

This might just be a style preference, but you could simplify this by directly calling `h.Rate.ServeHTTP(rw, req)`

File: internal/subscription/handler_test.go [hunk 1]
```diff
@@ -2,52 +2,72 @@ package subscription
...
 
-			subMock := new(SubscriberMock)
-			tt.mockFunc(subMock)
+			sub := tt.subscriberBuilder(ctrl)
 
-			h := NewHandler(subMock, log)
+			h := NewHandler(sub, log)
 
 			rw := httptest.NewRecorder()
 			handler := http.HandlerFunc(h.Subscribe)
```
Comment (lumenalux) [inline internal/subscription/handler_test.go:73]: Lines `72-73`

Is this line missing here `handler.ServeHTTP(rw, req)`?

File: internal/subscription/repository_test.go [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+			defer teardown()
+
+			repo := NewRepo(store)
+
+			var err error
+			for _, item := range tt.got {
+				err = repo.Add(item)
+				if err != nil {
+					break
+				}
```
Comment (lumenalux) [inline internal/subscription/repository_test.go:52]: If the function repo.Add(item) is capable of producing side effects, it may be desirable to ensure its execution for each item, even if an error occurs during the process. On the other hand, if encountering an error renders the subsequent execution of the loop unnecessary, it would be more advantageous to return immediately rather than simply breaking out of the loop

File: internal/subscription/repository_test.go
Comment (lumenalux) [inline internal/subscription/repository_test.go:1]: The repository tests cover only 'happy' and 'single failure' paths. Consider adding more complex scenarios, with several calls to `Add` and `GetAll` in a single test, possibly in random order

File: internal/subscription/subscription_test.go [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+			emailSenderBuilder: func(ctrl *gomock.Controller) EmailSender { return NewMockEmailSender(ctrl) },
+			wantErr:            fmt.Errorf("adding email subscription: %w", ErrEmailAlreadyExists),
+		},
+		// TODO(): Failed subscription due to internal server error.
+	}
+
+	for name, tc := range tests {
+		t.Run(name, func(t *testing.T) {
+			ctrl := gomock.NewController(t)
+			defer ctrl.Finish()
```
Comment (lumenalux) [inline internal/subscription/subscription_test.go:58]: `gomock.NewController` call and its corresponding `Finish` method are being repeated in every test case. So far, there are only two of tests, but there may be more in the future. Consider using `t.Cleanup` to register cleanup actions within each test case, which can make the code DRYer

File: internal/subscription/subscription_test.go [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+					Get().
+					Return(0.0, fmt.Errorf("failed to get rate")).
+					Times(1)
+				return mock
+			},
+			emailSenderBuilder: func(ctrl *gomock.Controller) EmailSender {
+				mock := NewMockEmailSender(ctrl)
+				mock.EXPECT().
+					Send(gomock.Any(), gomock.Any()).
+					Times(0)
```
Comment (lumenalux) [inline internal/subscription/subscription_test.go:139]: If there is no expectation for a method of a mock to be called during a test, there is no need to set up an expectation for that particular method

File: internal/subscription/subscription_test.go [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+			wantErr: nil,
+		},
+		"Failed to get rate": {
+			emails: []Email{{Address: &mail.Address{Address: "test1@example.com"}}, {Address: &mail.Address{Address: "test2@example.com"}}},
+			rate:   1.0,
+			repoBuilder: func(ctrl *gomock.Controller) EmailRepository {
+				mock := NewMockEmailRepository(ctrl)
+				mock.EXPECT().
+					GetAll().
+					Times(0)
```
Comment (lumenalux) [inline internal/subscription/subscription_test.go:124]: If there is no expectation for a method of a mock to be called during a test, there is no need to set up an expectation for that particular method

File: internal/subscription/subscription_test.go (hunk 4 above)
Comment (lumenalux) [inline internal/subscription/subscription_test.go:139]: If there is no expectation for a method of a mock to be called during a test, there is no need to set up an expectation for that particular method

File: internal/subscription/subscription_test.go [hunk 6]
```diff
@@ -0,0 +1,167 @@
...
+			emailSender := tc.emailSenderBuilder(ctrl)
+
+			svc := NewService(repo, rateGetter, emailSender)
+
+			err := svc.SendEmails()
+
+			if tc.wantErr != nil {
+				require.Error(t, err)
+				require.Equal(t, tc.wantErr.Error(), err.Error())
+				return
```
Comment (lumenalux) [inline internal/subscription/subscription_test.go:162]: Instead of comparing error messages to check if the returned error matches the expected error by string, it is advisable to use `errors.Is` to compare the error types

File: internal/subscription/subscription_test.go (hunk 5 above)
Comment (lumenalux) [inline internal/subscription/subscription_test.go:124]: It is unnecessary to set up an expectation for the `GetAll` method in the "Failed to get rate" test case. If a method of a mock is not expected to be called during a test, there is no need to explicitly define an expectation for it

File: internal/subscription/subscription_test.go [hunk 7]
```diff
@@ -0,0 +1,167 @@
...
+					Return(1.0, nil).
+					Times(1)
+				return mock
+			},
+			emailSenderBuilder: func(ctrl *gomock.Controller) EmailSender {
+				mock := NewMockEmailSender(ctrl)
+				mock.EXPECT().
+					Send(gomock.Any(), gomock.Any()).
+					Return(nil).
+					Times(2)
```
Comment (lumenalux) [inline internal/subscription/subscription_test.go:112]: The expectation for the `Send` method of the `EmailSender` is not specifying the specific parameters it should be called with. It is preferable to be explicit in setting expectations. Instead of using `gomock.Any()`, it is recommended to provide the exact parameters you expect the method to be called with

File: sys/filestore/filestore_test.go
Comment (lumenalux) [inline sys/filestore/filestore_test.go:1]: Line `55`

```Go
assert.Equal(t, tt.wantErr, err)
```

Consider using `errors.Is()` to check for specific error types instead of directly comparing errors

Comment (lumenalux) [inline sys/filestore/filestore_test.go:1]: Line `109`

```Go
assert.Equal(t, tt.want, fetchedItems)
```

This assertion assumes that the order of `fetchedItems` will be the same as the order of `tt.want`. If the implementation of `store.FetchAll()` does not guarantee the order of the returned items, this test may fail randomly

**Review by lumenalux — COMMENTED** [review]
Reviewed tests

File: internal/rate/handler_test.go
Comment (delveper) [reply internal/rate/handler_test.go:1]: Not exactly. `func(ResponseWriter, *Request)` does not implement `http.Handler` interface `ServeHTTP(w ResponseWriter, r *Request)`, but `http.HandlerFunc` does. I have to do type conversion, and there is a specific code in the standard library that allows this trick: 
```go
type HandlerFunc func(ResponseWriter, *Request)

// ServeHTTP calls f(w, r).
func (f HandlerFunc) ServeHTTP(w ResponseWriter, r *Request) {
	f(w, r)
}
```

File: internal/subscription/handler_test.go (hunk 1 above)
Comment (lumenalux) [reply internal/subscription/handler_test.go:72]: Resolved. My mistake. I did not expand the file view

**Review by lumenalux — APPROVED** [review]
Most of my comments have been corrected. The rest were more about the code style, and not about correctness of work or logic. So, good job. Approve

File: internal/subscription/handler.go [hunk 8]
```diff
@@ -84,6 +82,8 @@ func (h *Handler) Subscribe(rw http.ResponseWriter, req *http.Request) {
 		// TODO(not_documented): Handle server error.
 		h.log.Errorw("Writing response", "error", err)
 	}
+
+	h.log.Infow("Subscription successful.")
```
Comment (kachan1208) [inline internal/subscription/handler.go:86]: you don't need dots in the end of the logs.

examples:
 - https://github.com/uber-go/zap/blob/master/zaptest/logger_test.go look for examples/tests
 - consider checking STD library

File: internal/subscription/handler_test.go [hunk 9]
```diff
@@ -2,52 +2,71 @@ package subscription
...
-	"net/url"
 	"testing"
+
+	"github.com/GenesisEducationKyiv/main-project-delveper/sys/logger"
+	"github.com/golang/mock/gomock"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
 )
 
+//go:generate mockgen -destination=subscriber_mock_test.go -package=subscription github.com/GenesisEducationKyiv/main-project-delveper/internal/subscription Subscriber
```
Comment (kachan1208) [inline internal/subscription/handler_test.go:16]: such things should be in the places where interface is defined imagine changing an interface placement, go generate should migrate with it.

File: internal/subscription/handler_test.go [hunk 10]
```diff
@@ -2,52 +2,71 @@ package subscription
...
+
 func TestHandlerSubscribe(t *testing.T) {
```
Comment (kachan1208) [inline internal/subscription/handler_test.go:18]: consider adding more test cases

File: internal/subscription/repository_test.go [hunk 11]
```diff
@@ -0,0 +1,134 @@
...
+	"net/mail"
+	"os"
+	"testing"
+
+	"github.com/GenesisEducationKyiv/main-project-delveper/sys/filestore"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+)
+
+func TestRepoIntegration(t *testing.T) {
```
Comment (kachan1208) [inline internal/subscription/repository_test.go:13]: Such kind of tests is much simpler to organise not in table test form. It's hard to track what is going on here and which cases are covered.

File: internal/subscription/repository_test.go [hunk 12]
```diff
@@ -0,0 +1,134 @@
...
+	var err error
+	for _, email := range want {
+		err = repo.Store(email)
+		if err != nil {
+			break
+		}
+	}
+
+	if wantErr != nil {
+		assert.ErrorIs(t, err, wantErr)
```
Comment (kachan1208) [inline internal/subscription/repository_test.go:114]: consider changing asserts to requires.

File: internal/subscription/repository_test.go [hunk 13]
```diff
@@ -0,0 +1,134 @@
...
+	assert.NoError(t, err)
+}
+
+func testGetAll(t *testing.T, repo *Repo, wantErr error, want ...Email) {
+	t.Helper()
+
+	got, err := repo.GetAll()
+	if wantErr != nil {
+		assert.ErrorIs(t, err, wantErr)
+		require.Nil(t, got)
```
Comment (kachan1208) [inline internal/subscription/repository_test.go:127]: mixed usage of assert and require. consider changing asserts to requires.

File: internal/subscription/sender.go [hunk 14]
```diff
@@ -1,49 +1,49 @@
...
+
+// Send sends an email using the provided email address and rate.
+func (s *Sender) Send(email Email, rate float64) error {
+	subject := "Current BTC to UAH rate"
+
+	from := mail.NewEmail("Example Use", s.address)
+	to := mail.NewEmail(email.Address.Name, email.Address.String())
+
+	textContent := "Current rate is:"
+	htmlContent := fmt.Sprintf("<strong>%f</strong>", rate)
```
Comment (kachan1208) [inline internal/subscription/sender.go:33]: Next time for such things is much easier to use https://pkg.go.dev/text/template

It will make your life easier

File: internal/subscription/sender.go [hunk 15]
```diff
@@ -1,49 +1,49 @@
...
+
+	message := mail.NewSingleEmail(from, subject, to, textContent, htmlContent)
+
+	client := sendgrid.NewSendClient(s.apiKey)
```
Comment (kachan1208) [inline internal/subscription/sender.go:37]: do you need a client each time you want to send a message?
Can't you initialise client in the constructor method?

File: internal/subscription/sender.go [hunk 16]
```diff
@@ -1,49 +1,49 @@
...
+
+	resp, err := client.Send(message)
+	if err != nil {
+		return fmt.Errorf("sending email: %v", err)
+	}
+
+	if resp.StatusCode >= http.StatusBadRequest {
```
Comment (kachan1208) [inline internal/subscription/sender.go:44]: how do you divide 4xx codes from 5xx codes?

Usually 5xx means there is a critical problem and it should be logged and collected in metrics system and such kind of requests may be retried.

With 4xx in most cases you can't do anything as it's just a validation problem and you can log it as warning.

File: internal/subscription/subscription_test.go [hunk 17]
```diff
@@ -0,0 +1,171 @@
...
+
+	"github.com/golang/mock/gomock"
+	"github.com/stretchr/testify/require"
+)
+
+//go:generate mockgen -destination=email_repository_mock_test.go -package=subscription github.com/GenesisEducationKyiv/main-project-delveper/internal/subscription EmailRepository
+//go:generate mockgen -destination=rate_getter_mock_test.go -package=subscription github.com/GenesisEducationKyiv/main-project-delveper/internal/subscription RateGetter
+//go:generate mockgen -destination=email_sender_mock_test.go -package=subscription github.com/GenesisEducationKyiv/main-project-delveper/internal/subscription EmailSender
+
+func TestServiceSubscribe(t *testing.T) {
```
Comment (kachan1208) [inline internal/subscription/subscription_test.go:16]: - more test cases.
- consider moving go generate into places where interfaces are defined.
- split table tests into individual tests for simplicity. In current case you create constraints which limits you from custom tests.

File: internal/subscription/subscription_test.go [hunk 18]
```diff
@@ -0,0 +1,171 @@
...
+				mock := NewMockEmailRepository(ctrl)
+				mock.EXPECT().
+					Add(Email{Address: &mail.Address{Address: "test@example.com"}}).
+					Return(ErrEmailAlreadyExists).
+					Times(1)
+				return mock
+			},
+			rateGetterBuilder:  func(ctrl *gomock.Controller) RateGetter { return NewMockRateGetter(ctrl) },
+			emailSenderBuilder: func(ctrl *gomock.Controller) EmailSender { return NewMockEmailSender(ctrl) },
+			wantErr:            fmt.Errorf("adding email subscription: %w", ErrEmailAlreadyExists),
```
Comment (kachan1208) [inline internal/subscription/subscription_test.go:50]: you can check with `errors.Is` that `ErrEmailAlreadyExists` is inside your error?

File: internal/subscription/subscription_test.go (hunk 5 above)
Comment (kachan1208) [reply internal/subscription/subscription_test.go:129]: if you don't need a mock, just don't write it. YAGNI principle.

What can be more explicit than absence of mock?)
Imagine also having more than 10 mocks that you don't need, why to define them?

File: internal/subscription/subscription_test.go (hunk 6 above)
Comment (kachan1208) [reply internal/subscription/subscription_test.go:166]: STD lib doesn't mean an external dependency it's a functionality that is available from go 1.13.

File: internal/subscription/subscription_test.go [hunk 19]
```diff
@@ -0,0 +1,171 @@
...
+
+			repo := tc.repoBuilder(ctrl)
+			rateGetter := tc.rateGetterBuilder(ctrl)
+			emailSender := tc.emailSenderBuilder(ctrl)
+
+			svc := NewService(repo, rateGetter, emailSender)
+
+			err := svc.SendEmails()
+
+			if tc.wantErr != nil {
```
Comment (kachan1208) [inline internal/subscription/subscription_test.go:163]: do you need this check?

Can't we just write:
```
				require.Equal(t, tc.wantErr, err)
```

if it's err is nil and want is nil it will be true.

File: sys/filestore/filestore.go [hunk 20]
```diff
@@ -134,3 +145,35 @@ func (f *FileStore[T]) FetchAll() ([]T, error) {
 
 	return coll, nil
 }
+
+func (f *FileStore[T]) StoreAll(items ...T) error {
```
Comment (kachan1208) [inline sys/filestore/filestore.go:149]: woo generics.

File: sys/filestore/filestore.go [hunk 21]
```diff
@@ -134,3 +145,35 @@ func (f *FileStore[T]) FetchAll() ([]T, error) {
...
+	return hex.EncodeToString(h.Sum(nil))
+}
+
+func TestSetup[T any](t *testing.T) (*FileStore[T], func()) {
+	t.Helper()
+
+	dir := t.TempDir()
+	store := New[T](dir)
+
+	teardown := func() {
```
Comment (kachan1208) [inline sys/filestore/filestore.go:172]: you don't need this, use: https://pkg.go.dev/testing/fstest

File: test/Dockerfile [hunk 22]
```diff
@@ -0,0 +1,4 @@
+FROM node:lts-alpine3.17
+RUN npm install -g newman
+WORKDIR /app
+CMD ["newman", "run", "postman.json"]
```
Comment (kachan1208) [inline test/Dockerfile:4]: why postman/newman, why not to start application and write another go application that performs request?

File: internal/subscription/subscription_test.go (hunk 6 above)
Comment (delveper) [reply internal/subscription/subscription_test.go:166]: @kachan1208 I would like to make myself clear because `errors.Is` is not a dependency. The errors from other packages interfaces that we are mocking still are dependencies.

File: internal/rate/handler.go [hunk 23]
```diff
@@ -3,13 +3,12 @@ package rate
...
 	"encoding/json"
-	"github.com/GenesisEducationKyiv/main-project-delveper/sys/logger"
 	"net/http"
+
+	"github.com/GenesisEducationKyiv/main-project-delveper/sys/logger"
 )
 
 // Getter interface to get rate from external service.
-//
-//go:generate moq -out getter_mock_test.go . Getter
```
[the remark this answers is not in the document]
Comment (delveper) [reply internal/rate/handler.go:12]: @kachan1208 Not really. I switched from moq to gomock and wrote all needful `go:generate` inside test files.
```go

//go:generate mockgen -destination=email_repository_mock_test.go -package=subscription github.com/GenesisEducationKyiv/main-project-delveper/internal/subscription EmailRepository
//go:generate mockgen -destination=rate_getter_mock_test.go -package=subscription github.com/GenesisEducationKyiv/main-project-delveper/internal/subscription RateGetter
//go:generate mockgen -destination=email_sender_mock_test.go -package=subscription github.com/GenesisEducationKyiv/main-project-delveper/internal/subscription EmailSender

```

File: internal/rate/handler_test.go [hunk 24]
```diff
@@ -3,31 +3,45 @@ package rate
...
 	"net/http/httptest"
 	"testing"
+
+	"github.com/GenesisEducationKyiv/main-project-delveper/sys/logger"
+	"github.com/golang/mock/gomock"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
 )
 
+//go:generate mockgen -destination=getter_mock_test.go -package=rate github.com/GenesisEducationKyiv/main-project-delveper/internal/rate Getter
```
[the remark this answers is not in the document]
Comment (delveper) [reply internal/rate/handler_test.go:17]: @kachan1208 Nice to hear that, but I was advised not to use `moq` because `gomock` is "standard". Looks like opinions vary. BTW I am agree with you about `moq` and this Ryer looks like a cool guy 😎

File: internal/rate/handler_test.go [hunk 25]
```diff
@@ -3,31 +3,45 @@ package rate
...
 		"Rate retrieval failure": {
-			mockFunc: func(m *GetterMock) {
-				m.GetFunc = func() (float64, error) {
-					return 0, errors.New("failed to retrieve rate")
-				}
+			getterBuilder: func(ctrl *gomock.Controller) Getter {
+				mock := NewMockGetter(ctrl)
+				mock.EXPECT().
+					Get().
+					Return(0.0, errors.New("failed to retrieve rate")).
```
[the remark this answers is not in the document]
Comment (delveper) [reply internal/rate/handler_test.go:42]: @kachan1208 Sure it will work. It is weird, but explicit. 
Чистий код pdf - Google Drive - Google Chrome

File: internal/rate/handler_test.go [hunk 26]
```diff
@@ -3,31 +3,45 @@ package rate
...
 
+//go:generate mockgen -destination=getter_mock_test.go -package=rate github.com/GenesisEducationKyiv/main-project-delveper/internal/rate Getter
+
 func TestHandlerRate(t *testing.T) {
-	cases := map[string]struct {
-		mockFunc func(m *GetterMock)
-		wantCode int
-		want     any
+	tests := map[string]struct {
+		getterBuilder func(*gomock.Controller) Getter
```
[the remark this answers is not in the document]
Comment (delveper) [reply internal/rate/handler_test.go:21]: @kachan1208 I agree with you. `moq` will address the issue with ugliness. However, closures mocks and a lack of structured table tests could also lead in some places to a big ball of mud. It seems like this is a matter of personal preferences. Please correct me if I am wrong.

File: internal/rate/handler_test.go (hunk 25 above)
[the remark this answers is not in the document]
Comment (delveper) [reply internal/rate/handler_test.go:42]: I am a victim of copy/paste 😸

File: sys/filestore/filestore.go (hunk 20 above)
Comment (delveper) [reply sys/filestore/filestore.go:149]: Generics have been available in Go since version 1.18, and they are not going anywhere 😎

File: test/Dockerfile (hunk 22 above)
Comment (delveper) [reply test/Dockerfile:4]: @kachan1208 I do not mind doing that, but will make a Postman collection anyway. Why don't we use it this way? Is this approach not valid for end-to-end testing?

File: internal/rate/handler_test.go (hunk 24 above)
[the remark this answers is not in the document]
Comment (kachan1208) [reply internal/rate/handler_test.go:17]: gomock is not standard, you will not find it in some projects.
It was archived since yesterday and screenshot proves that it can't be a standard at least since yesterday)

[the remark this answers is not in the document]
Comment (delveper) [reply internal/rate/handler_test.go:17]: @kachan1208 Thank you for rapid answer. Is it mandatory to get rid of it? Because it is not abandoned, but forked to https://github.com/uber/mock

File: internal/rate/handler_test.go [hunk 27]
```diff
@@ -3,31 +3,45 @@ package rate
...
+		want          any
 	}{
 		"Valid rate": {
-			mockFunc: func(m *GetterMock) { m.GetFunc = func() (float64, error) { return 2.5, nil } },
+			getterBuilder: func(ctrl *gomock.Controller) Getter {
+				mock := NewMockGetter(ctrl)
+				mock.EXPECT().
+					Get().
+					Return(2.5, nil).
+					Times(1)
```
[the remark this answers is not in the document]
Comment (delveper) [reply internal/rate/handler_test.go:31]: I will check it twice. I got some related errors here, that force me to use explicit calls everywhere

File: internal/subscription/sender.go (hunk 14 above)
Comment (delveper) [reply internal/subscription/sender.go:33]: There is still much room for improvement.

File: internal/rate/handler_test.go (hunk 24 above)
[the remark this answers is not in the document]
Comment (kachan1208) [reply internal/rate/handler_test.go:17]: not mandatory. 
I don't request most changes, I'm helping you to understand a different perspective of the question. In some below comments there are no suggestions to change something, but recommendations.

File: internal/subscription/sender.go (hunk 16 above)
Comment (delveper) [reply internal/subscription/sender.go:44]: Noted

File: internal/subscription/subscription_test.go (hunk 17 above)
Comment (delveper) [reply internal/subscription/subscription_test.go:16]: Thank you for your useful recommendations. I am aware of the need to add more test cases, but our API is really simple so far and I have only implemented the bare minimum according to the initial requirements (I am going to replace docs with a detailed openapi). To add more valid test cases, I need to change the source code dramatically. I am going to do this by introducing a new way of doing handlers and then adding more test cases. To be honest, I struggled with test tables in some places and will avoid adding them where it is not convenient.

File: internal/subscription/handler.go (hunk 8 above)
Comment (delveper) [reply internal/subscription/handler.go:86]: Whether or not to use punctuation in log messages is a matter of convention. We have only conventions related to the punctuation of errors in Go community, right? Some projects prefer to use complete sentences with punctuation, as this makes the messages easier to read and understand. Other projects prefer to use shorter messages with no punctuation, as this makes them easier to scan quickly. Ultimately, the best approach depends on the specific needs of the project. BTW I wonder why stdlib logger sucks.

File: sys/filestore/filestore.go (hunk 21 above)
Comment (delveper) [reply sys/filestore/filestore.go:172]: I do not get it. Could you please explain how integrate it in my test?

File: internal/subscription/handler.go (hunk 8 above)
Comment (kachan1208) [reply internal/subscription/handler.go]: Yep. I just relay on go source code in this case.
go source code for logging: https://github.com/search?q=repo%3Agolang%2Fgo+log.Print&type=code&p=2
e.g for errors(just as additional reading): https://gist.github.com/adamveld12/c0d9f0d5f0e1fba1e551#error-strings

File: internal/subscription/sender.go (hunk 16 above)
Comment (kachan1208) [reply internal/subscription/sender.go]: but you can detect anomalies of them in metrics later)

File: test/Dockerfile (hunk 22 above)
Comment (kachan1208) [reply test/Dockerfile:4]: you can use it, I just asked a question to understand an intention)

File: internal/subscription/subscription_test.go (hunk 5 above)
Comment (delveper) [reply internal/subscription/subscription_test.go]: Fixed

File: internal/rate/handler_test.go [hunk 28]
```diff
@@ -36,14 +50,15 @@ func TestHandlerRate(t *testing.T) {
...
 
-	for name, tt := range cases {
+	for name, tt := range tests {
 		t.Run(name, func(t *testing.T) {
-			req, _ := http.NewRequest(http.MethodGet, "/api/rate", nil)
+			ctrl := gomock.NewController(t)
 
-			getterMock := new(GetterMock)
-			tt.mockFunc(getterMock)
+			req, err := http.NewRequest(http.MethodGet, "/api/rate", nil)
```
[the remark this answers is not in the document]
Comment (delveper) [reply internal/rate/handler_test.go:46]: The code was rearranged following the instructions

File: internal/subscription/email_sender_mock_test.go [hunk 29]
```diff
@@ -0,0 +1,48 @@
+// Code generated by MockGen. DO NOT EDIT.
```
[the remark this answers is not in the document]
Comment (delveper) [reply internal/subscription/email_sender_mock_test.go]: Done

File: internal/subscription/repository_test.go (hunk 11 above)
Comment (delveper) [reply internal/subscription/repository_test.go:15]: Done

File: internal/subscription/handler_test.go (hunk 10 above)
Comment (delveper) [reply internal/subscription/handler_test.go:16]: Done

File: internal/subscription/repository_test.go (hunk 12 above)
Comment (delveper) [reply internal/subscription/repository_test.go]: Done

**Review by kachan1208 — APPROVED** [review]
